### ISSUE TITLE
Coder unnecessarily replaces EC2 instance on workspace start

### ISSUE THREAD

**Issue by mjswensen** [coder/coder#5615]
# Steps to reproduce

1. Spin up a new workspace from the template below
2. Do some work in the workspace
3. Stop the workspace (or let it stop automatically)
4. Note in the AWS Console that the EC2 instance associated with the workspace is in a "Stopped" state (not "Terminated").
5. Start the workspace again from the Coder UI

## Expected outcome

I expected the stopped instance to simply be started, with the work saved to its volume in step 2 above persisted.

## Observed outcome

The underlying Terraform process reports "drift detected," so the stopped instance is terminated (and the work contained in its volume is lost), a new instance (and new attached volume) is created in its place.

# Template file

My template file is based heavily on [the official AWS Linux example](https://github.com/coder/coder/blob/main/examples/templates/aws-linux/main.tf):

```tf
terraform {
  required_providers {
    coder = {
      source  = "coder/coder"
      version = "0.6.6"
    }
    aws = {
      source  = "hashicorp/aws"
      version = "4.23.0"
    }
  }
}

provider "aws" {
  region = "us-east-1"
}

resource "coder_agent" "main" {
  arch = "amd64"
  os   = "linux"
  auth = "aws-instance-identity"
}

data "coder_workspace" "me" {}

locals {

  linux_user = "coder"

  user_data_start = <<EOT
Content-Type: multipart/mixed; boundary="//"
MIME-Version: 1.0

--//
Content-Type: text/cloud-config; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="cloud-config.txt"

#cloud-config
cloud_final_modules:
- [scripts-user, always]
hostname: ${lower(data.coder_workspace.me.name)}
users:
  - default
  - name: ${local.linux_user}
    sudo: ALL=(ALL) NOPASSWD:ALL
    shell: /bin/bash

--//
Content-Type: text/x-shellscript; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="userdata.txt"

#!/bin/bash
sudo -u ${local.linux_user} sh -c '${coder_agent.main.init_script}'
--//--
EOT

  user_data_end = <<EOT
Content-Type: multipart/mixed; boundary="//"
MIME-Version: 1.0

--//
Content-Type: text/cloud-config; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="cloud-config.txt"

#cloud-config
cloud_final_modules:
- [scripts-user, always]

--//
Content-Type: text/x-shellscript; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="userdata.txt"

#!/bin/bash
sudo shutdown -h now
--//--
EOT

}

resource "aws_instance" "eng" {
  ami                         = "ami-06878d265978313ca"
  availability_zone           = "us-east-1a"
  instance_type               = "t3.xlarge"
  subnet_id                   = "<REDACTED>"
  associate_public_ip_address = true
  key_name                    = "<REDACTED>"
  vpc_security_group_ids = [
    "<REDACTED>",
    "<REDACTED>",
    "<REDACTED>",
  ]
  root_block_device {
    volume_size = 128
  }
  user_data = data.coder_workspace.me.transition == "start" ? local.user_data_start : local.user_data_end
  tags = {
    Name              = "coder-${data.coder_workspace.me.id}-${data.coder_workspace.me.name}"
    Coder_Provisioned = "true"
  }
}

resource "coder_metadata" "workspace_info" {
  resource_id = aws_instance.eng.id
  item {
    key   = "private IP address"
    value = aws_instance.eng.private_ip
  }
}
```

Here is the relevant snippet from the "Planning infrastructure" step when Coder starts the workspace:

```
data.coder_workspace.me: Refreshing...
coder_agent.main: Refreshing state... [id=3dc150f8-b3f3-4434-8b48-dcedcadb510a]
data.coder_workspace.me: Refresh complete after 0s [id=627a7a27-4b54-4c1e-8e13-b626ddc43810]
coder_agent.main: Refresh complete [id=3dc150f8-b3f3-4434-8b48-dcedcadb510a]
aws_instance.eng: Refreshing state... [id=i-07a2c151ebe104fb3]
aws_instance.eng: Refresh complete [id=i-07a2c151ebe104fb3]
coder_metadata.workspace_info: Refreshing state... [id=a768600c-9e62-4ae4-9fb1-3d953e713dc0]
coder_metadata.workspace_info: Refresh complete [id=a768600c-9e62-4ae4-9fb1-3d953e713dc0]
aws_instance.eng: Drift detected (update)
coder_agent.main: Drift detected (update)
aws_instance.eng: Plan to replace
coder_metadata.workspace_info: Plan to replace
Plan: 2 to add, 0 to change, 2 to destroy.
```

Without the ability to see the underlying Terraform diff/output, I'm not sure why there is "drift detected" (which is what is forcing Terraform to do a replacement).

**Comment by mattlqx** [comment]
For me, this occurred when the data source I was using to get the latest AMI would change. Looking at your resource, this would appear to be because you're using the user data to do different things depending on Coder's lifecycle, that attribute change is very likely the reason the instance resource is tainted and would be recreated.

If you want to keep using user data in that way, you can just add a `lifecycle` block to ignore changes to the user data attribute and prevent that from triggering a recreate.

```
resource "aws_instance" "eng" {
  lifecycle {
    ignore_changes = [
      user_data,
    ]
  }
  ...
```

**Comment by mjswensen** [comment]
Thanks @mattlqx , I really appreciate you taking the time to write that up. It makes a lot of sense. Though the `null_resource` is a great trick, I'm still getting a full replacement of the instance upon workspace restart; Terraform is reporting `aws_instance.eng: Plan to replace`.

<details>
<summary>Here's the new template with @mattlqx 's changes.</summary>

```tf
terraform {
  required_providers {
    coder = {
      source  = "coder/coder"
      version = "0.6.6"
    }
    aws = {
      source  = "hashicorp/aws"
      version = "4.23.0"
    }
  }
}

provider "aws" {
  region = "us-east-1"
}

resource "coder_agent" "main" {
  arch = "amd64"
  os   = "linux"
  auth = "aws-instance-identity"
}

data "coder_workspace" "me" {}

resource "aws_instance" "eng" {
  ami                         = "ami-06878d265978313ca"
  availability_zone           = "us-east-1a"
  instance_type               = "t3.xlarge"
  subnet_id                   = "<REDACTED>"
  associate_public_ip_address = true
  key_name                    = "<REDACTED>"
  vpc_security_group_ids = [
    "<REDACTED>",
    "<REDACTED>",
    "<REDACTED>",
  ]
  root_block_device {
    volume_size = 128
  }
  user_data = <<EOT
Content-Type: multipart/mixed; boundary="//"
MIME-Version: 1.0

--//
Content-Type: text/cloud-config; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="cloud-config.txt"

#cloud-config
cloud_final_modules:
- [scripts-user, always]
hostname: ${lower(data.coder_workspace.me.name)}
users:
  - default
  - name: coder
    sudo: ALL=(ALL) NOPASSWD:ALL
    shell: /bin/bash

--//
Content-Type: text/x-shellscript; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="userdata.txt"

#!/bin/bash
sudo -u coder sh -c '${coder_agent.main.init_script}'
--//--
EOT

  tags = {
    Name              = "coder-${data.coder_workspace.me.owner}-${data.coder_workspace.me.name}"
    Coder_Provisioned = "true"
  }
}

resource "null_resource" "sleep" {
  count = data.coder_workspace.me.start_count == 0 ? 1 : 0

  triggers = {
    instance_id = aws_instance.eng.id
  }

  provisioner "local-exec" {
    command = "aws ec2 stop-instances --instance-ids ${aws_instance.eng.id}"
  }

  provisioner "local-exec" {
    when    = destroy
    command = "aws ec2 start-instances --instance-ids ${self.triggers.instance_id}"
  }
}
```

</details>

Now that the conditional user_data is removed, I can't see what might still be causing the instance to be tainted.

**Comment by mjswensen** [comment]
(Coder team, aside from any pointers on the template itself—which are greatly appreciated—if there could be a way to see a detailed version of the Terraform plan that would be immensely helpful in debugging templates!)

**Comment by mattlqx** [comment]
Glad to help. I don't see what could be triggering a recreation given that config, everything seems like it should be static.

I agree the Terraform run could be more descriptive by default, none of that output is from Coder though. There may be some opportunity to debug if you set [TF_LOG](https://developer.hashicorp.com/terraform/internals/debugging) to debug in whatever env you run the coder server from (systemd unit or whereever) to get more output and it's usually a LOT of output.

**Comment by mjswensen** [comment]
Another great suggestion—but unfortunately Coder doesn't seem to be affected by the `TF_LOG` environment variable (nor `TF_LOG_PATH`). 🤔

**Comment by mattlqx** [comment]
How are you setting it? I'm running Coder through systemd and setting `TF_LOG=debug` in `/etc/coder.d/coder.env` (assuming you're using their unit) definitely does get through to Terraform.

<img width="1421" alt="Screenshot 2023-01-08 at 8 38 14 AM" src="https://user-images.githubusercontent.com/739851/211205725-4cd7018b-9c96-42d8-ba2d-4eaad3fd5536.png">

**Comment by mjswensen** [comment]
Oops, I had set it in `/etc/environment` and run `sudo systemctl restart coder` (yes, I used their default install script) thinking it could pick it up from there. Didn't think to put it in `coder.env` 🤦‍♂️ 

Looking at the logs, the most likely culprits are side effects of shutting down the instance:

```
2023-01-09T14:13:38.883Z [WARN]  Provider "registry.terraform.io/hashicorp/aws" produced an unexpected new value for aws_instance.eng during refresh.
      - .public_ip: was cty.StringVal("44.202.137.90"), but now cty.StringVal("")
      - .associate_public_ip_address: was cty.True, but now cty.False
      - .instance_state: was cty.StringVal("running"), but now cty.StringVal("stopped")
      - .public_dns: was cty.StringVal("ec2-44-202-137-90.compute-1.amazonaws.com"), but now cty.StringVal("")
```

(Interestingly, it does still mention changes to `user_data` as well, which makes me wonder if there is something dynamic hiding in `coder_agent.main.init_script`... doesn't seem like it on first glance 🤷‍♂️ )

I went ahead and added this to the instance configuration:

```
lifecycle {
  ignore_changes = all
}
```

Which I think should work well enough for this particular template.

Thank you again for all your help, @mattlqx !

**Comment by mjswensen** [comment]
To close the loop for Coder engineering:

- [ ] Consider documenting more clearly that the [AWS Linux example](https://github.com/coder/coder/blob/main/examples/templates/aws-linux/main.tf) destroys and re-creates the instance during the workspace's stop/start cycle.
- [ ] Consider documenting @mattlqx 's approach to instance shut down as an alternate example (or replacement to the current)—in my opinion it is more likely to line up with what users are expecting.
    - The official example is actually a little odd—when stopping the workspace, the instance shuts down, gets the new user_data (which is simply `sudo shutdown -h now`), boots up, and then runs the user data which shuts it back down. It's a lot of runaround for something that's just going to get destroyed anyway the next time the workspace starts.
- [ ] I also noticed that the [permissions policy provided in the AWS Linux example](https://github.com/coder/coder/tree/main/examples/templates/aws-linux#required-permissions--policy) were insufficient to spin up the example workspace. I replaced them with `ec2:*` to work around this (unfortunately, AWS doesn't seem to have an quick/easy way to identify which permissions are needed for a particular request, so I didn't investigate much farther than that).
- [ ] Consider adding to the documentation a section on tips for debugging templates—and include adding `TF_LOG` to `/etc/coder.d/coder.env` for folks to see more Terraform output when necessary.

I'll leave this ticket open so you all can decide what you'd like to do, if anything. I'd be happy to PR some documentation changes as well if you all like a particular suggestion. Feel free to close if not. 🙂 Thanks for Coder, it's a fantastic tool and I'm looking forward to trying it out with the team now that we're up and running!
